Modules/binascii.c

```diff
@@ -1239,13 +1239,101 @@ binascii_b2a_qp_impl(PyObject *module, Py_buffer *data, int quotetabs,
     return rv;
 }
 
+/*[clinic input]
+binascii.b2a_base85
+
+    data: Py_buffer
+    chars: Py_buffer
+    pad: bool = False
+    foldnuls: bool = False
+    foldspaces: bool = False
+
+Utility method used by the base64 module to encode a85/b85 data
+
+    data: bytes
+    chars: 85 bytes conversion table
+    pad: use NULL-paded input if necessary
+    foldnuls: replace NULL chunks by 'z'
+    foldspaces: replace space-only chucks by 'y'
+
+[clinic start generated code]*/
+
+static PyObject *
+binascii_b2a_base85_impl(PyObject *module, Py_buffer *data, Py_buffer *chars,
+                         int pad, int foldnuls, int foldspaces)
+/*[clinic end generated code: output=0a92b3c535580aa0 input=a2d8ae712ed5adba]*/
+{
+    if (chars->len != 85) {
+        PyErr_SetString(PyExc_ValueError,
+                        "chars must be exactly 85 bytes long");
+        return NULL;
+    }
+
+    _PyBytesWriter writer;
+    _PyBytesWriter_Init(&writer);
+
+    const size_t bin_len = data->len;
+
+    // Allocate up to maxium encoded length, adjusted at end
+    const size_t ascii_len = ((bin_len + 3) / 4) * 5;
+
+    unsigned char *ascii_data = _PyBytesWriter_Alloc(&writer, ascii_len);
+    if (ascii_data == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    const unsigned char *table = chars->buf;
+    const unsigned char *bin_data = data->buf;
+
+    size_t i, j;
+    for (i = 0; i < bin_len; i += 4) {
+        const size_t chunk_size = (bin_len - i >= 4) ? 4 : (bin_len - i);
+
+        // translate chunk to 32bit integer
+        uint32_t value = 0;
+        for (j = 0; j < chunk_size; j++) {
+            value = (value << 8) | bin_data[i + j];
+        }
+        value <<= (4 - chunk_size) * 8;
+
+        if (foldnuls && value == 0) {
+            *ascii_data++ = 'z';
+        } else if (foldspaces && value == 0x20202020) {
+            *ascii_data++ = 'y';
+        } else {
+            for (j = 0; j < 5 ; j++) {
+                ascii_data[4 - j] = table[value % 85];
+                value /= 85;
+            }
+            ascii_data += 5;
+        }
+    }
+
+    // In case `i` went over the input size, we may need to shorten the output
+    const size_t overflow = (i - bin_len);
+
+    if (overflow && !pad && foldnuls && ascii_data[-1] == 'z') {
```

Comment:
`ascii_data[-1]` won't work in C :') I think you meant `ascii_data[strlen(ascii_data) - 1]`. Alternatively I'd suggest keeping an accumulator variable that remembers the number of characters that were written.

Reply:
I was as surprised as you, but it seems to work (since I got a regression where this condition was triggered when it should not have been)

From what I understand, `ascii_data` is a pointer that is used to "write" to the buffer allocated by the writer, and its position is the next byte that the loop will write to.

So accessing `ascii_data[0]` is the byte at the current position `ascii_data[1]` the next one, and `ascii_data[-1]` the previous one.

Unless I'm abusing some part of the C standard I do not understand

Reply:
No, you're right. it's actually correct per §6.5.2.1/2 since it's equivalent to `+ offset` (and the offset here is -1). You just need to guarantee that it points to valid memory index. Now the only way to trigger it is if ascii_data is empty, which is only the cased when bin_len = 0 and then `overflow` would be false so no way to trigger it.